xdsl/transforms/csl_stencil_handle_async_flow.py

```diff
@@ -0,0 +1,233 @@
+            for a in op.iter_args
+        )
+
+        no_params = FunctionType.from_lists([], [])
+        cond_task_id = self.counter + 1
+
+        pre_block = op.parent_block()
+        if pre_block is None:
+            return
+        post_block = pre_block.split_before(op)
+
+        post_func = csl.FuncOp(f"for_post{self.counter}", no_params)
+        rewriter.inline_block(post_block, InsertPoint.at_start(post_func.body.block))
+
+        cond_func = csl.TaskOp(
+            f"for_cond{self.counter}",
+            no_params,
+            task_kind=csl.TaskKind.LOCAL,
+            id=cond_task_id,
+        )
+        body_func = csl.FuncOp(f"for_body{self.counter}", no_params)
+        inc_func = csl.FuncOp(f"for_inc{self.counter}", no_params)
+
+        # create csl.vars for loop var and iter_args outside the parent func
+        rewriter.insert_op(
+            iv := csl.VariableOp.from_value(IntegerAttr(op.lb.op.value.value, 16)),
+            InsertPoint.before(parent_func),
+        )
+        iter_vars = [csl.VariableOp.from_type(arg_t) for arg_t in op.iter_args.types]
+        rewriter.insert_op(iter_vars, InsertPoint.before(parent_func))
+
+        # parent func (pre loop): setup iter vars and activate cond_func
+        with ImplicitBuilder(pre_block):
+            for dst, src in zip(iter_vars, op.iter_args):
+                csl.StoreVarOp(dst, src)
+            csl.ActivateOp(cond_task_id, csl.TaskKind.LOCAL)
+            # csl.CallOp(SymbolRefAttr(cond_func.sym_name))
+            csl.ReturnOp()
+
+        # for-loop condition func
+        with ImplicitBuilder(cond_func.body.block):
+            ub = arith.Constant.from_int_and_width(op.ub.op.value.value, 16)
+            iv_load = csl.LoadVarOp(iv)
+            cond = arith.Cmpi(iv_load, ub, "slt")
+            branch = scf.If(cond, [], Region(Block()), Region(Block()))
+            with ImplicitBuilder(branch.true_region):
+                csl.CallOp(SymbolRefAttr(body_func.sym_name))
+                scf.Yield()
+            with ImplicitBuilder(branch.false_region):
+                csl.CallOp(SymbolRefAttr(post_func.sym_name))
+                scf.Yield()
+            csl.ReturnOp()
+
+        # for-loop inc func
+        with ImplicitBuilder(inc_func.body.block):
+            step = arith.Constant.from_int_and_width(op.step.op.value.value, 16)
+            iv_load = csl.LoadVarOp(iv)
+            stepped = arith.Addi(iv_load, step)
+            csl.StoreVarOp(iv, stepped)
+            load_vars = [csl.LoadVarOp(v) for v in iter_vars]
+            for iter_var, yielded_var in zip(iter_vars, terminator.arguments):
+                idx = (
+                    op.body.block.args.index(yielded_var) - 1
+                )  # subtract 1 as the first block arg is the loop var
+                csl.StoreVarOp(iter_var, load_vars[idx])
+            csl.ActivateOp(cond_task_id, csl.TaskKind.LOCAL)
+            # csl.CallOp(SymbolRefAttr(cond_func.sym_name))
```

Comment:
Again, a tiny bit of information why you choose to use task instead of a func would be appreciated.

Also, commented out code should probably be deleted.